#11 feat make oauth signinsignup feature

uridonget opened the pull request:
# 42 OAuth를 통한 회원관리 및 JWT 발급 \#11
## Overview
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

User 테이블의 username 기본값을 UUID로 변경
- JWT 인증 과정에서 username이 사용자와 직접적으로 연관된 데이터를 포함할 경우, 보안상 위험 요소가 될 가능성 존재함. 
- 예측 불가능한 UUID를 username 기본값으로 설정하여 보안을 강화.

42 OAuth Authorization Code 발급을 위한 리다이렉트 함수 추가
- 프론트엔드에서 42 OAuth Authorization URI로 직접 리다이렉트하는 방식은 URI가 사용자에게 노출되어 보안상 문제가 발생할 수 있음.
- 서버에 요청을 보내 Authorization URI로 리다이렉트하도록 구현. 이를 통해 민감한 URI 정보가 클라이언트 측에서 노출되지 않고 서버 측에서 안전하게 처리됨.

42 OAuth Callback 함수 추가
- OAuth 인증 과정을 통해 42 서버로부터 유저 데이터를 받아옴
- 유저 데이터를 기반으로 사용자의 쿠키에 JWT를 저장하는 로직을 추가
- 유저가 데이터베이스에 등록되지 않은(회원가입 하지 않은) 경우에는 새로운 유저 데이터를 생성
- 쿠키를 발급하고 사용자를 기본페이지로 리다이렉트 시킴

<!---- Resolves: #(Isuue Number) -->
## PR Type
어떤 변경 사항이 있나요?

- [ ] feat

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

uridonget committed:
#11 feat(OAuth): Authorization 코드 발급

42 OAuth API를 생성.
주어지는 CLIENT_ID, REDIRECT_URL을 환경변수로 등록.
추가한 환경변수를 사용해 Authorization 코드를 발급 받을 수 있는 42 로그인 URL을 생성.
해당 URL으로 redirect.

uridonget committed:
#11 feat(OAuth): 42 OAuth를 통한 JWT 발급 기능 추가

42 OAuth를 통한 JWT 발급을 구현하였습니다.

1. User 테이블의 username 기본값을 수정할 수 없는 UUID로 설정
    - 위치 : backend/account/models.py

2. simplejwt 라이브러리 설치 추가
    - 위치 : backend/requirements.txt
    - djangorestframework-simplejwt 추가

3. simplejwt 관련 설정 추가
    - 위치 : backend/transcendence/settings.py
    - ACCESS/REFRESH 토큰의 유효 기간 등 설정 추가

4. 42 OAuth 인증 콜백 처리 함수(callback) 추가
    - 위치 : backend/account/views.py
    - 인증 코드를 통해 access_token 및 refresh_token을 요청
    - 42 API를 호출하여 사용자 정보(인트라 ID, 이메일, 프로필 이미지)를 조회
    - OAuth 테이블을 기반으로 사용자 정보를 생성하거나 조회
    - 생성된 JWT를 클라이언트의 쿠키에 저장 후 기본 페이지(http://localhost)로 리디렉션

5. 42 OAuth callback request에 따라 callback 함수를 실행하는 엔드포인트 추가
    - 위치 : backend/account/urls.py

6. 랜덤 닉네임 생성 함수(generate_random_nickname) 추가
    - 위치 : backend/account/views.py
    - USER# + 랜덤 숫자 형식의 닉네임 생성
    - 중복 확인 후 닉네임 리턴

7. 유저 조회 및 생성 함수(get_or_create_user_oauth) 추가
    - 위치 : backend/account/views.py
    - OAuth 테이블을 통해 인트라 ID 기준으로 사용자 조회
    - 기존 사용자가 없으면 새로운 유저와 OAuth 객체 생성

8. docker-compose.yml의 네트워크 설정 변경
    - 위치 : docker-compose.yml
    - frontend 와 동일한 네트워크를 사용하기 위해 네트워크를 external 설정으로 변경

allorak333 commented on backend/account/views.py:
```diff
@@ -1,4 +1,173 @@
+from account.models import OAuth
+
+
+@api_view(["GET"])
+def get_oauth_token(request):
+    """
+    @brief 클라이언트를 42 API의 OAuth 인증 페이지로 리디렉션하는 함수
+
+    @param request Django의 HTTP 요청 객체
+
+    @return 42 API의 OAuth 인증 페이지로 리디렉션
+
+    @details
+    환경 변수에서 CLIENT_ID와 REDIRECT_URI를 읽어와 OAuth 인증 URL을 생성합니다.
+    생성된 URL로 리디렉션하여 42 인증을 시작합니다.
+    """
+
+    client_id = os.environ.get("CLIENT_ID")
+    redirect_uri = os.environ.get("REDIRECT_URI")
+    authorize_uri = f"https://api.intra.42.fr/oauth/authorize?client_id={client_id}&redirect_uri={redirect_uri}&response_type=code&scope=public"
+    return redirect(authorize_uri)
```
해당 api의 결과로 데이터베이스에 유저 정보가 삽입될 수 있기에 GET보다는 POST가 더 적절하다고 생각합니다!

uridonget replied:
저도 동의합니다. 제가 저렇게 GET으로 설정한 이유는 OAuth 인증의 특성상 페이지 이동이 필요하기 때문에 프론트에서 redirect를 사용할 것 같다고 생각해서였습니다. POST method를 사용하면서 페이지 이동도 할 수 있는 방법이 있다면 당연히 수정의사가 있습니다. 이 건은 수요일 정기 회의에서 대화를 나눠보는 것이 좋을 것 같습니다!

allorak333 commented on backend/transcendence/settings.py:
```diff
@@ -39,9 +39,30 @@
+SIMPLE_JWT = {
+    "ACCESS_TOKEN_LIFETIME": timedelta(minutes=15),
+    "REFRESH_TOKEN_LIFETIME": timedelta(days=1),
+    "ROTATE_REFRESH_TOKENS": False,
+    "BLACKLIST_AFTER_ROTATION": True,
```
ROTATE_REFRESH_TOKEN을 true로 설정한 후 BLACKLIST_AFTER_ROTATION을 true로 설정했을 경우 refresh token으로 access token를 발급 후 기존 refresh token은 블랙리스트에 추가됩니다.

ROTATE_REFRESH_TOKENS가 False, BLACKLIST_AFTER_ROTATION 가 True인 경우에는 어떻게 동작하나요?

uridonget replied:
ROTATE_REFRESH_TOKENS가 False, BLACKLIST_AFTER_ROTATION 가 True인 경우에는 블랙리스트에 어떤 refresh 토큰도 등록되지 않을 것 입니다. 사실상 설정이 아무 의미가 없어진다고 보면 됩니다.
보안을 생각할 때 ROTATE_REFRESH_TOKENS 설정을 True로 두는 것도 고려해보겠습니다. 토큰을 refresh 하는 부분은 아직 코드를 작성하지 않았기 때문에 추후 해당 부분을 작성할 때 커밋 자세하게 적도록 하겠습니다

uridonget committed:
#11 fix(OAuth): get_oauth_token 메소드 제한 수정 및 불필요한 설정 제거

get_oauth_token 함수의 허용 메소드를 GET에서 모든 메소드 허용으로 변경.
settings.py에서 simplejwt와 관련된 불필요한 설정 제거.

allorak333 reviewed (approved):
LGTM!

allorak333 commented on backend/account/views.py:
```diff
@@ -1,4 +1,172 @@
+        user, created = get_or_create_user_oauth(intra_id, user_email, user_image_path)
+        refresh = RefreshToken.for_user(user)
+        access = refresh.access_token
+        response = HttpResponseRedirect("http://localhost")
```
이 부분은 하드 코딩보다는 환경 변수로 남겨두는 편이 좋을 것 같습니다.
만약에 프론트에서 로그인 페이지의 경로를 변경한다거나, 주소값이 달라지는 경우 수정이 번거로울 수 있어요.

allorak333 commented on backend/account/views.py:
```diff
@@ -1,4 +1,172 @@
+        intra_id = user_info.get("login")
+        user_email = user_info.get("email")
+        user_image_path = user_info.get("image", {}).get("link")
+        user, created = get_or_create_user_oauth(intra_id, user_email, user_image_path)
```
`get_or_create_user_oauth` 함수 내부에 exception을 throw하는 로직이 있지만 그것을 catch하는 로직이 존재하지 않습니다. 예외 처리에 대한 로직이 필요해보입니다.
